### PR TITLE
sql,descs: return clear error when failing due to session expiration

### PR DESCRIPTION
**Pull request by ajwerner**
First commit refactors `TestTransactionDeadline` a bit to make the dependency
injection a bit more direct and the logic a bit more compact. It changes the
testing knob to now only override optionally as opposed to always.

Before this commit, when a sql query failed because its session was expired,
the connection would disconnect without any error. This commit fixes that
error handling and provides a clearer error message indicating the source
of the problem.
    
Release note: None

### CODE REVIEW

**Review by fqazi — COMMENTED** [review]
> Reviewed 4 of 4 files at r1, 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70657)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @rimadeodhar)

**Review by rimadeodhar — APPROVED** [review]
Thanks for the fix. :lgtm: 

> Reviewed 1 of 2 files at r2, 2 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70657#-:-MkNf1ft9cuvEfeO7FNr:b-y5b93v)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @fqazi and @rimadeodhar)

**Comment by ajwerner** [comment]
TFTR!

bors r+

File: pkg/sql/conn_executor_test.go [hunk 1]
```diff
@@ -1054,26 +1051,44 @@ func TestTransactionDeadline(t *testing.T) {
...
+	sessionOverrideKnob := func(ctx context.Context) (sqlliveness.Session, error) {
+		return sessionOverride.Load().(sessionOverrideFunc)(ctx)
+	}
+	sessionOverride.Store(noopSessionOverrideFunc)
+	// setClientSessionOverride takes a session to return in the case that
+	// the context is due to a client connection. It returns a closure to
+	// reset the session override hook.
+	setClientSessionOverride := func(fs sqlliveness.Session) (reset func()) {
+		sessionOverride.Store(func(ctx context.Context) (sqlliveness.Session, error) {
+			// This somewhat hacky approach allows the hook to itentify contexts
```
Comment (Azhng) [inline pkg/sql/conn_executor_test.go:1070]: super nit: s/itentify/identiy

**Pull request merged**